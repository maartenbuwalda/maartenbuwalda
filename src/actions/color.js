export const SET_DOMINANT_COLOR = 'SET_DOMINANT_COLOR'
export const GET_DOMINANT_COLOR = 'GET_DOMINANT_COLOR'

// export const setDominantColor = palette => {
//   return {
//     type: SET_DOMINANT_COLOR,
//     payload: {
//       colorRegular: palette.color,
//       colorLight: palette.colorLight,
//       colorDark: palette.colorDark,
//     }
//   }
// }

export const getDominantColor = image => {
  return {
    type: GET_DOMINANT_COLOR,
    payload: {
      image
    }
  }
}
