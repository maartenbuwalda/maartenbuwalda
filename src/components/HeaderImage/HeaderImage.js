import React from 'react'
import styled from 'styled-components'
import Img from 'react-image'
import PropTypes from 'prop-types'
import Loading from '../Loading'
import { sizes } from '../../constants/mediaQueries'
import colors from '../../constants/colors'

class HeaderImage extends React.Component {
  componentDidMount () {
    const { updateTheme, image } = this.props
    updateTheme(image)
  }

  render () {
    const { image } = this.props

    return (
      <ImageWrapper>
        <Img
          src={image}
          loader={<Placeholder><Loading/></Placeholder>}/>
      </ImageWrapper>
    )
  }
}

const ImageWrapper = styled.figure`
  background: ${colors.gray};
  margin: 0;
  overflow: hidden;
  height: 30rem;
  max-height: 20rem;
  width: 100%;
  position: relative;
  display: flex;
  justify-content: center;
  align-items: center;

  img {
    position: absolute;
    animation: appear .5s ease-out;
    display: block;
    max-height: 100%;
  }

  @media (min-width: ${sizes.m}) {
    max-height: 30rem;
  }
`

const Placeholder = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  min-height: 30rem;
  width: 100%;
  background: ${colors.gray};
`

HeaderImage.propTypes = {
  image: PropTypes.string,
  updateTheme: PropTypes.func,
}

export default HeaderImage
