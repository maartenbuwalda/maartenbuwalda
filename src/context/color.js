import React from 'react'

export const colors = {
  hex: {
    colorMutedRegular: '',
    colorMutedLight: '',
    colorMutedDark: '',
    colorRegular: '',
    colorLight: '',
    colorDark: '',
  },
  rgb: {
    colorMutedRegular: '',
    colorMutedLight: '',
    colorMutedDark: '',
    colorRegular: '',
    colorLight: '',
    colorDark: '',
  }
}

export const ColorContext = React.createContext({
  theme: colors,
  updateTheme: () => {},
})
