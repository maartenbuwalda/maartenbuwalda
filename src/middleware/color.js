import {
  GET_DOMINANT_COLOR,
  SET_DOMINANT_COLOR,
} from '../actions/color'

import Vibrant from 'node-vibrant'

const color = store => next => action => {
  next(action)
  const dispatch = store.dispatch

  if (action.type === GET_DOMINANT_COLOR) {
    const img = new Vibrant(action.payload.image)
    img.getPalette((err, palette) => {
      if (!err) {
        dispatch({
          type: SET_DOMINANT_COLOR,
          payload: {
            hex: {
              colorMutedDark: palette.DarkMuted && palette.DarkMuted.getHex(),
              colorMutedLight: palette.LightMuted && palette.LightMuted.getHex(),
              colorMutedRegular: palette.Muted && palette.Muted.getHex(),
              colorDark: palette.DarkVibrant && palette.DarkVibrant.getHex(),
              colorLight: palette.LightVibrant && palette.LightVibrant.getHex(),
              colorRegular: palette.Vibrant && palette.Vibrant.getHex(),
            },
            rgb: {
              colorMutedDark: palette.DarkMuted && palette.DarkMuted.getRgb().toString(),
              colorMutedLight: palette.LightMuted && palette.LightMuted.getRgb().toString(),
              colorMutedRegular: palette.Muted && palette.Muted.getRgb().toString(),
              colorDark: palette.DarkVibrant && palette.DarkVibrant.getRgb().toString(),
              colorLight: palette.LightVibrant && palette.LightVibrant.getRgb().toString(),
              colorRegular: palette.Vibrant && palette.Vibrant.getRgb().toString(),
            }
          }
        })
      }
    })
  }
}

export default color
