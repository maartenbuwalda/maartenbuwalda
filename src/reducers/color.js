import {
  SET_DOMINANT_COLOR,
  GET_DOMINANT_COLOR,
} from '../actions/color'

const defaultState = {
  isFetching: false,
  hex: {
    colorMutedRegular: '',
    colorMutedLight: '',
    colorMutedDark: '',
    colorRegular: '',
    colorLight: '',
    colorDark: '',
  },
  rgb: {
    colorMutedRegular: '',
    colorMutedLight: '',
    colorMutedDark: '',
    colorRegular: '',
    colorLight: '',
    colorDark: '',
  }
}

function colorReducer (state = defaultState, action) {
  switch (action.type) {
    case GET_DOMINANT_COLOR:
      return {
        ...state,
        isFetching: true,
      }
    case SET_DOMINANT_COLOR:
      return {
        ...state,
        isFetching: false,
        hex: action.payload.hex,
        rgb: action.payload.rgb,
      }
    default:
      return state
  }
}

export { colorReducer }
