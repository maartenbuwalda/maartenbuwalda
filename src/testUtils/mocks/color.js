export const color = {
  hex: {
    colorMutedRegular: '',
    colorMutedLight: '',
    colorMutedDark: '',
    colorRegular: '',
    colorLight: '',
    colorDark: '',
  },
  rgb: {
    colorMutedRegular: '',
    colorMutedLight: '',
    colorMutedDark: '',
    colorRegular: '',
    colorLight: '',
    colorDark: '',
  }
}
