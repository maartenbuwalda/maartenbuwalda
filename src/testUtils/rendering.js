import toJson from 'enzyme-to-json'
import { shallow, mount } from 'enzyme'
import configureStore from 'redux-mock-store'
import 'jest-styled-components'

export const componentRenders = component => {
  it('renders correctly', () => {
    const shallowed = shallow(component)
    expect(toJson(shallowed)).toMatchSnapshot()
  });
}

export const connectedComponentRenders = component => {
  const middlewares = []
  const mockStore = configureStore(middlewares)
  const initialState = {}
  const store = mockStore(initialState)
  
  it('renders correctly', () => {
    const shallowed = shallow(component, store)
    expect(toJson(shallowed)).toMatchSnapshot()
  })
}
