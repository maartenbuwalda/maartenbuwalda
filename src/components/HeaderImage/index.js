import HeaderImage from './HeaderImage'
import { withRouter } from 'react-router'

export default withRouter(HeaderImage)
