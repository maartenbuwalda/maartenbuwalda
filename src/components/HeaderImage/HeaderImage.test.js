import React from 'react'
import HeaderImage from './HeaderImage'
import { componentRenders } from '../../testUtils/rendering'
import { color } from '../../testUtils/mocks/color'

describe('HeaderImage component', () => {
  const testTitle = 'testTitle'
  const component = <HeaderImage title={testTitle} theme={color} updateTheme={jest.fn()} />
  componentRenders(component)
})
