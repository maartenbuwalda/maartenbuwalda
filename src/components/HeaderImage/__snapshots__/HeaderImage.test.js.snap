// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderImage component mounts correctly 1`] = `
.c0 {
  background: #E5E5E8;
  margin: 0;
  overflow: hidden;
  height: 30rem;
  max-height: 15rem;
  width: 100%;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 img {
  -webkit-animation: appear .5s ease-out;
  animation: appear .5s ease-out;
  display: block;
  max-height: 100%;
}

@media (min-width:768px) {
  .c0 {
    max-height: 30rem;
  }
}

<HeaderImage
  theme={
    Object {
      "hex": Object {
        "colorDark": "colorDark",
        "colorLight": "colorLight",
        "colorMutedDark": "colorMutedDark",
        "colorMutedLight": "colorMutedLight",
        "colorMutedRegular": "colorMutedRegular",
        "colorRegular": "colorRegular",
      },
      "rgb": Object {
        "colorDark": "colorDark",
        "colorLight": "colorLight",
        "colorMutedDark": "colorMutedDark",
        "colorMutedLight": "colorMutedLight",
        "colorMutedRegular": "colorMutedRegular",
        "colorRegular": "colorRegular",
      },
    }
  }
  title="testTitle"
  updateTheme={
    [MockFunction] {
      "calls": Array [
        Array [
          undefined,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  <styled.figure>
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-htpNat",
            "isStatic": true,
            "lastClassName": "c0",
            "rules": Array [
              "
  background: ",
              "#E5E5E8",
              ";
  margin: 0;
  overflow: hidden;
  height: 30rem;
  max-height: 15rem;
  width: 100%;
  position: relative;
  display: flex;
  justify-content: center;
  align-items: center;

  img {
    animation: appear .5s ease-out;
    display: block;
    max-height: 100%;
  }

  @media (min-width: ",
              "768px",
              ") {
    max-height: 30rem;
  }
",
            ],
          },
          "displayName": "styled.figure",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-htpNat",
          "target": "figure",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <figure
        className="c0"
      >
        <Img
          container={[Function]}
          decode={true}
          loader={
            <ForwardRef>
              <Loading />
            </ForwardRef>
          }
          src={Array []}
          unloader={false}
        />
      </figure>
    </StyledComponent>
  </styled.figure>
</HeaderImage>
`;
